File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.py
from collections import deque

def solution(rectangle, characterX, characterY, itemX, itemY):
    SIZE = 102
    board = [[0] * SIZE for _ in range(SIZE)]

    # 1. 좌표 2배 확장해서 사각형 채우기
    for x1, y1, x2, y2 in rectangle:
        x1 *= 2; y1 *= 2; x2 *= 2; y2 *= 2
        for x in range(x1, x2 + 1):
            for y in range(y1, y2 + 1):
                board[y][x] = 1

    # 2. 내부 비우기 (1로 채운 후 진짜 내부를 -1로)
    for x1, y1, x2, y2 in rectangle:
        x1 *= 2; y1 *= 2; x2 *= 2; y2 *= 2
        for x in range(x1 + 1, x2):
            for y in range(y1 + 1, y2):
                board[y][x] = -1

    # 3. BFS 탐색
    visited = [[0] * SIZE for _ in range(SIZE)]
    q = deque()
    sx, sy = characterX * 2, characterY * 2
    ex, ey = itemX * 2, itemY * 2
    q.append((sy, sx))
    visited[sy][sx] = 1

    dx = [1, -1, 0, 0]
    dy = [0, 0, 1, -1]

    while q:
        y, x = q.popleft()
        if x == ex and y == ey:
            return (visited[y][x] - 1) // 2  # 2배 좌표였으므로 절반

        for i in range(4):
            nx, ny = x + dx[i], y + dy[i]
            if 0 <= nx < SIZE and 0 <= ny < SIZE:
                if board[ny][nx] == 1 and visited[ny][nx] == 0:
                    visited[ny][nx] = visited[y][x] + 1
                    q.append((ny, nx))

File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.js
class Queue {
    constructor() {
        this.store = {};
        this.front = 0;
        this.rear = 0;
    }
    
    size() {
        if (this.store[this.rear] === undefined) {
            return 0;
        } else {
            return this.rear - this.front + 1;
        }
    }
    
    push(data) {
        if (this.size() === 0) {
            this.store[this.rear] = data;
        } else {
            this.rear += 1;
            this.store[this.rear] = data;
        }
    }
    
    popLeft() {
        let popData = this.store[this.front];
        delete this.store[this.front];
        
        if (this.front === this.rear) {
            this.front = 0;
            this.rear = 0;
        } else {
            this.front += 1;
        }
        
        return popData;
    }
    
    peekFront() {
        if (this.size() === 0) {
            return undefined;
        } else {
            return this.store[this.front];
        }
    }
    
    peekRear() {
        if (this.size() === 0) {
            return undefined;
        } else {
            return this.store[this.rear];
        }
    }
}

function solution(rectangle, characterX, characterY, itemX, itemY) {
    const boardSize = 102;
    const board = Array.from({ length: boardSize }, () => Array(boardSize).fill(0));
    
    for (const r of rectangle) {
        const x1 = r[0] * 2;
        const y1 = r[1] * 2;
        const x2 = r[2] * 2;
        const y2 = r[3] * 2;
        
        for (let i = x1; i < x2 + 1; i++) {
            for (let j = y1; j < y2 + 1; j++) {
                board[i][j] = 1;
            }
        }
    }
    
    for (const r of rectangle) {
        const x1 = r[0] * 2;
        const y1 = r[1] * 2;
        const x2 = r[2] * 2;
        const y2 = r[3] * 2;
        
        for (let i = x1 + 1; i < x2; i++) {
            for (let j = y1 + 1; j < y2; j++) {
                board[i][j] = -1;
            }
        }
    }
    
    const startX = characterX * 2;
    const startY = characterY * 2;
    const endX = itemX * 2;
    const endY = itemY * 2;
    
    const direction = [[1, 0], [0, 1], [-1, 0], [0, -1]];
    const visited = Array.from({ length: boardSize }, () => Array(boardSize).fill(0));
    const queue = new Queue();
    
    visited[startX][startY] = 1;
    queue.push([startX, startY]);
    
    while (queue.size() > 0) {
        const current = queue.popLeft();
        const currentX = current[0];
        const currentY = current[1];
        
        if (currentX === endX && currentY === endY) {
            return Math.floor((visited[currentX][currentY] - 1) / 2);
        }
        
        for (const d of direction) {
            const nextX = currentX + d[0];
            const nextY = currentY + d[1];
            
            if (nextX >= 0 && nextX < 102 && nextY >= 0 && nextY < 102) {
                if (visited[nextX][nextY] === 0 && board[nextX][nextY] === 1) {
                    queue.push([nextX, nextY]);
                    visited[nextX][nextY] = visited[currentX][currentY] + 1;
                }
            }
        }
    }
}